## ひろしまQuest2020#stayhome【コース予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_12.f'
SUBMIT = 'submit/course_2.csv'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 171)

### 不要な列を削除

In [4]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [5]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [6]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 156)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  bc_straight  \
0           0           0        0         0           0     0.477023   
1           0           0        0         0           1     0.459832   
2           0           0        0         0           2     0.469730   
3           0           0        1         0           0     0.477023   
4           0           0        1         0           1     0.459832   

   bc_curve  bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  \
0  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
1  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   
2  0.066062   0.159395  0.031361  0.162940     0.062722   0.014112   
3  0.097592   0.200870  0.074454  0.046799     0.041882   0.008619   
4  0.090366   0.161893  0.068313  0.099045     0.061079   0.012505   

   bc_cutball  bc_area0  bc_area1  bc_area2  bc_area3  bc_area4  bc_area5  \
0    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
1    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   
2    0.033679  0.021884  0.035656  0.030884  0.020316  0.021748  0.021816   
3    0.052760  0.048489  0.071105  0.059259  0.047675  0.063792  0.049150   
4    0.046967  0.035008  0.056450  0.047706  0.035812  0.044074  0.034687   

   bc_area6  bc_area7  bc_area8  bc_area9  bc_area10  bc_area11  bc_area12  \
0  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
1  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   
2  0.028429  0.036678  0.029520  0.133761   0.131170   0.201595   0.286542   
3  0.046338  0.070075  0.059013  0.117934   0.090325   0.114800   0.162045   
4  0.037484  0.058379  0.050407  0.112451   0.103867   0.161700   0.221976   

   first  second  third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0       0      0         0       0.0     1.0        1     1        0   
1      0       0      0         0       0.0     1.0        1     1        0   
2      0       0      0         0       0.0     1.0        1     1        0   
3      0       0      0         0       0.0     0.0        1     1        0   
4      0       0      0         0       0.0     0.0        1     1        0   

   date_from_opening  ...  firm_bat  height_bat  weight_bat  draft_order_bat  \
0                  0  ...         0         176          74              5.0   
1                  0  ...         0         176          74              5.0   
2                  0  ...         0         176          74              5.0   
3                  0  ...         0         180          74              1.0   
4                  0  ...

### train
- 行数: 257117

In [7]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 156)

### test
- 行数: 521650

In [8]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 156)

In [9]:
del all_pitch
gc.collect()

20

In [10]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [11]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [12]:
lgb_train = lgb.Dataset(train_d, train['pitch_area'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [13]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 13,
        'seed' : 0,
        'learning_rate' : 0.1,
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 2.34432 + 0.000533242
[200]	cv_agg's multi_logloss: 2.34533 + 0.00072475
Best num_boost_round: 121
Best CV score: 2.344104226715559
CPU times: user 47min 37s, sys: 22.5 s, total: 47min 59s
Wall time: 12min 30s


In [14]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [15]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 10min 22s, sys: 4.25 s, total: 10min 26s
Wall time: 2min 42s


### Feature Importance

In [16]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
19                bc_shoot       606
85            changeup_pit       606
100        salary_year_pit       615
147         salary_dif_b-c       632
89        pit_game_cnt_pit       638
81               curve_pit       639
151  salary_x_year_dif_b-c       642
102                bmi_pit       647
110       bat_game_cnt_bat       665
118                bmi_bat       684
77              height_pit       694
146            bmi_dif_p-c       698
83               shoot_pit       704
16             bc_straight       710
52              point_diff       718
1        pitch_cnt_in_game       733
138    salary_year_dif_p-b       760
104             height_bat       769
150    salary_year_dif_b-c       810
84                fork_pit       812
105             weight_bat       819
82              slider_pit       925
0               No_in_game       962
152            bmi_dif_b-c       966
140            bmi_dif_p-b      1008
109         batter_cnt_bat      1018
47             elapsed_min      1135
8    pitcher_cnt_in_inning      1207
7      pitcher_cnt_in_game      1433
46       date_from_opening      1802

In [17]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 2min 58s, sys: 2.52 s, total: 3min
Wall time: 50.2 s


## 結果出力

In [18]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.065832  0.141877  0.082695  0.049623  0.066723  0.036926   
1      1  0.067888  0.149608  0.066369  0.040060  0.060172  0.041020   
2      2  0.033871  0.081491  0.051842  0.028368  0.039711  0.029953   
3      3  0.045628  0.101019  0.056334  0.029763  0.046285  0.033365   
4      4  0.049009  0.133540  0.057184  0.035096  0.049181  0.039911   

          6         7         8         9        10        11        12  
0  0.065967  0.049041  0.033646  0.158395  0.071418  0.080262  0.097596  
1  0.047208  0.075656  0.033142  0.153112  0.066502  0.102057  0.097206  
2  0.042157  0.048704  0.027861  0.231430  0.077416  0.182971  0.124226  
3  0.045215  0.048779  0.028743  0.210331  0.098375  0.125644  0.130518  
4  0.048771  0.060897  0.031547  0.189801  0.064672  0.121641  0.118751

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [19]:
submit.mean()

index    260824.500000
0             0.047909
1             0.068319
2             0.056601
3             0.047029
4             0.060693
5             0.047252
6             0.043046
7             0.065001
8             0.053944
9             0.109329
10            0.088372
11            0.133705
12            0.178798
dtype: float64

In [20]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [21]:
'signate submit --competition-id=276 ~/work/baseball/{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=276 ~/work/baseball/submit/course_2.csv --note feat=156_cv=2.344104226715559'

### GCS

In [22]:
'gsutil cp ~/work/baseball/{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ~/work/baseball/submit/course_2.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648